radsat-sk/operation/subsystems/RBattery.c

```diff
@@ -0,0 +1,279 @@
+
+/** Number of voltage/current calls to the Battery */
+#define NUMBER_OF_VOLTAGE_AND_CURRENT_COMMANDS (3)
+
+/** Number of temperature calls to the Battery */
+#define NUMBER_OF_TEMP_COMMANDS (4)
+
+
+/***************************************************************************************************
+                                          PRIVATE GLOBALS
+***************************************************************************************************/
+
+/**
+ * 3 Byte commands for requesting output voltage readings for each bus
+ */
+static uint32_t batteryVoltageCommandBytes[3] = {
+		0x10E280,		// Output Voltage of Battery
+		0x10E210,		// Output Voltage of 5V Bus
+		0x10E200		// Output Voltage of 3.3V Bus
+};
+
+/**
+ * 3 Byte commands for requesting output current readings for each bus
+ */
+static uint32_t batteryCurrentCommandBytes[3] = {
+		0x10E284,		// Output Current of Battery Bus in mA
+		0x10E214,		// Output Current of 5V Bus in mA
+		0x10E204		// Output Current of 3.3V Bus in mA
+};
+
+/**
+ * 4 Byte commands for requesting temperature readings for each board
+ */
+static uint32_t batteryTemperatureCommandBytes[4] = {
+		0x10E308,		// Motherboard Temperature
+		0x10E398,		// Daughterboard 1 Temperature
+		0x10E3A8,		// Daughterboard 2 Temperature
+		0x10E3B8		// Daughterboard 3 Temperature
+};
+
+/***************************************************************************************************
+                                       PRIVATE FUNCTION STUBS
+***************************************************************************************************/
+
+static int batteryTalk(uint8_t* command, uint8_t* response);
+
+
+/***************************************************************************************************
+                                             PUBLIC API
+***************************************************************************************************/
+
+/**
+ * Request and store all of the relavant data (voltage, current, temperature) from the battery board
+ * 		and battery daughter boards
+ * @return A BatteryStatus object containing all of the telemetry of the battery board
+ */
+int getBatteryTelemetry(BatteryStatus* dataStorage){
+
+	// Create temporary variables for sending and receiving I2C data
+	uint8_t i2c_command[BATTERY_TELEM_COMMAND_LENGTH] = {0};
+	uint8_t i2c_data[BATTERY_RESPONSE_LENGTH] = {0};
+
+	// Create a temporary array for calculated voltage and current before transferring into the BatteryStatus structure
+	float storedData[NUMBER_OF_VOLTAGE_AND_CURRENT_COMMANDS] = {0};
+
+	// Create a temporary array for calculated temperatures before transferring into the BatteryStatus structure
+	float storedTempData[NUMBER_OF_TEMP_COMMANDS] = {0};
+
+
+	// Send 3 commands to get ADC output voltage readings from the Battery
+	for(int i = 0; i < NUMBER_OF_VOLTAGE_AND_CURRENT_COMMANDS; i = i + 1){
+		memcpy(i2c_command, &batteryVoltageCommandBytes[i], BATTERY_TELEM_COMMAND_LENGTH);
```

Comment:
If re-using a buffer for different purposes, you should make a habit of completely clearing the buffer beforehand.

e.g. add `memset(i2c_command, 0, sizeof(i2c_command));` before every `memcpy` line

Reply:
Same goes for data, you should clear it before passing it to another function (since it gets set by the talk function)